### PR TITLE
Updated etag in memory repository behaviour 

### PR DESCRIPTION
**Pull request by robertbennett1998**
Updated the in memory repository to reflect the real behaviour when upserting items.

Added the ability to set the etag in the constructor this will allow the items to be mapped to and from other objects without losing this data.

# Breaking Changes
- Changed item with etag to be abstract

### CODE REVIEW

File: src/Microsoft.Azure.CosmosRepository/EtagItem.cs [hunk 1]
```diff
@@ -30,8 +30,25 @@ namespace Microsoft.Azure.CosmosRepository
...
     /// </example>
-    public class EtagItem : Item, IItemWithEtag
+    public abstract class EtagItem : Item, IItemWithEtag
     {
+        /// <summary>
+        /// Default constructor
+        /// </summary>
+        protected EtagItem()
+        {
+
```
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/EtagItem.cs:40]:
```suggestion
```

File: src/Microsoft.Azure.CosmosRepository/FullItem.cs [hunk 2]
```diff
@@ -34,6 +34,24 @@ namespace Microsoft.Azure.CosmosRepository
     /// </example>
     public abstract class FullItem : Item, IItemWithEtag, IItemWithTimeToLive, IItemWithTimeStamps
     {
+
+        /// <summary>
+        /// Default constructor
+        /// </summary>
+        protected FullItem()
+        {
+
```
Comment (IEvangelist) [inline src/Microsoft.Azure.CosmosRepository/FullItem.cs:43]:
```suggestion
```

**Review by IEvangelist — DISMISSED** [review]
Looks good to me. Thoughts @mumby0168

**Review by mumby0168 — APPROVED** [review]

**Comment by mumby0168** [comment]
> Looks good to me. Thoughts @mumby0168

Yeah looks good to me, some issues we found while trying these out for real. Nice work @robertbennett1998 thanks, again.

The breaking changes come at a good time since we are about to bump the major version. Was there anything in terms of breaking changes you wanted to add @IEvangelist ? I have some time this evening.

**Review by mumby0168 — APPROVED** [review]

**Comment by IEvangelist** [comment]
> > Looks good to me. Thoughts @mumby0168
> 
> Yeah looks good to me, some issues we found while trying these out for real. Nice work @robertbennett1998 thanks, again.
> 
> The breaking changes come at a good time since we are about to bump the major version. Was there anything in terms of breaking changes you wanted to add @IEvangelist ? I have some time this evening.

Nothing that comes to mind, @mumby0168

**Pull request merged**